src/OpenTelemetry.Contrib.Instrumentation.GrpcCore/ServerTracingInterceptor.cs

```diff
@@ -0,0 +1,211 @@
+                var requestStreamReaderProxy = new AsyncStreamReaderProxy<TRequest>(
+                    requestStream,
+                    rpcScope.RecordRequest);
+
+                var responseStreamProxy = new ServerStreamWriterProxy<TResponse>(
+                    responseStream,
+                    rpcScope.RecordResponse);
+
+                await continuation(requestStreamReaderProxy, responseStreamProxy, context).ConfigureAwait(false);
+                rpcScope.Complete();
+            }
+            catch (Exception e)
+            {
+                rpcScope.CompleteWithException(e);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// A class to help track the lifetime of a service-side RPC.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request.</typeparam>
+        /// <typeparam name="TResponse">The type of the response.</typeparam>
+        private class ServerRpcScope<TRequest, TResponse> : RpcScope<TRequest, TResponse>
+            where TRequest : class
+            where TResponse : class
+        {
+            /// <summary>
+            /// The metadata setter action.
+            /// </summary>
+            private static readonly Func<Metadata, string, IEnumerable<string>> MetadataGetter = (metadata, key) =>
+            {
+                for (var i = 0; i < metadata.Count; i++)
+                {
+                    var entry = metadata[i];
+                    if (entry.Key.Equals(key))
+                    {
+                        return new string[1] { entry.Value };
+                    }
+                }
+
+                return Enumerable.Empty<string>();
+            };
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="ServerRpcScope{TRequest, TResponse}"/> class.
+            /// </summary>
+            /// <param name="context">The context.</param>
+            /// <param name="options">The options.</param>
+            public ServerRpcScope(ServerCallContext context, ServerTracingInterceptorOptions options)
+                : base(context.Method, options.RecordMessageEvents)
+            {
+                if (!GrpcCoreInstrumentation.ActivitySource.HasListeners())
+                {
+                    return;
+                }
+
+                var currentContext = Activity.Current?.Context;
+
+                // Extract the SpanContext, if any from the headers
+                var metadata = context.RequestHeaders;
+                if (metadata != null)
+                {
+                    var propagationContext = options.Propagator.Extract(new PropagationContext(currentContext ?? default, Baggage.Current), metadata, MetadataGetter);
+                    if (propagationContext.ActivityContext.IsValid())
+                    {
+                        currentContext = propagationContext.ActivityContext;
+                        Baggage.Current = propagationContext.Baggage;

```

Comment:
I think most of the existing instrumentation will set baggage independently of the activity context, eg:
```csharp
if (ctx.ActivityContext.IsValid())
{
   // set activity context
}

if (ctx.Baggage != default)
{
   Baggage.Current = ctx.Baggage;
}
```
Not sure if there is a compelling reason to do that, I guess it is theoretically possible to have baggage extracted by itself?

Reply:
I confess I'm a bit ignorant about baggage. I've never used it. I'll add the default check prior to setting Baggage.Current